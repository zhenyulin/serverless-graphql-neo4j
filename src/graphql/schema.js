import { makeExecutableSchema } from 'apollo-server-lambda';
import { augmentSchema } from 'neo4j-graphql-js';

import typeDefs from './type';
import resolvers from './resolver';

export default augmentSchema(
	makeExecutableSchema({
		typeDefs,
		resolvers,
		resolverValidationOptions: {
			requireResolversForResolveType: false,
		},
	}),
);
