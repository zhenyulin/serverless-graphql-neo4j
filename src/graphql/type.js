export default `
	type User {
		id: ID!
		name: String
		email: String!
		follows: [User] @relation(name: "FOLLOWS", direction: "OUT")
		followers: [User] @relation(name: "FOLLOWS", direction: "IN")
		likes: [Item] @relation(name: "LIKES", direction: "OUT")
	}

	type Item {
		id: ID!
		name: String
		likedBy: [User] @relation(name: "LIKES", direction: "IN")
	}

	type Query {
		user(id: ID!): User
		item(id: ID!): Item
	}
`;
