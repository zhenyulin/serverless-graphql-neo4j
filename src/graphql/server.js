import express from 'express';
import serverless from 'serverless-http';
import expressGraphql from 'express-graphql';
import { makeExecutableSchema } from 'graphql-tools';

import typeDefs from './type';
import resolvers from './resolver';

const schema = makeExecutableSchema({ typeDefs, resolvers });
const handler = expressGraphql({
	schema,
	graphiql: true,
});
const app = express().use(handler);

export default serverless(app);
