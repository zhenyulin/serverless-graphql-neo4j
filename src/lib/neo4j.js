import { v1 as neo4j } from 'neo4j-driver';

const driver = neo4j.driver(
	process.env.NEO4J_URI,
	neo4j.auth.basic(process.env.NEO4J_USER, process.env.NEO4J_PASSWORD),
);

export const runCypher = async (cypher, param, debug) => {
	if (debug) console.log(cypher);
	const session = driver.session();
	try {
		const { records } = await session.run(cypher, param);
		const items = records.map(r => r.get(0).properties);
		if (debug) console.log(items);
		return items;
	} catch (e) {
		console.log(e);
		throw e;
	} finally {
		session.close();
	}
};

export const runCypherReturnOne = async (cypher, param, debug) => {
	if (debug) console.log(cypher);
	const session = driver.session();
	try {
		const { records } = await session.run(cypher, param);
		const items = records.map(r => r.get(0).properties);
		if (debug) console.log(items);
		return items[0];
	} catch (e) {
		if (
			e.code &&
			e.code === 'Neo.ClientError.Schema.ConstraintValidationFailed'
		) {
			e.message = 'Node with constraint properties already exists';
		}
		throw e;
	} finally {
		session.close();
	}
};

export const runTransactionWithArray = async (cypher, params, debug) => {
	const session = driver.session();
	const transaction = session.beginTransaction();
	try {
		const result = [];
		params.forEach(async param => {
			if (debug) console.log(cypher, param);
			const r = await transaction.run(cypher, param);
			result.push(r);
		});
		await transaction.commit();
		return result;
	} catch (e) {
		transaction.rollback();
		throw e;
	} finally {
		session.close();
	}
};

export const runCyphersAsTransactions = async cyphers => {
	const session = driver.session();
	const transaction = session.beginTransaction();
	try {
		const result = [];
		cyphers.forEach(async cypher => {
			const r = await transaction.run(cypher);
			result.push(r);
		});
		await transaction.commit();
		return result;
	} catch (e) {
		transaction.rollback();
		throw e;
	} finally {
		session.close();
	}
};

export default driver;
