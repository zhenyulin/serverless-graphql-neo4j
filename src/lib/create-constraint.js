import { runCypher } from './neo4j';

export const createUserConstraint = () =>
	runCypher(`
	CREATE CONSTRAINT ON (u:User) ASSERT u.email IS UNIQUE
`);

export const createItemConstraint = () =>
	runCypher(`
	CREATE CONSTRAINT ON (i:Item) ASSERT i.name IS UNIQUE
`);

export default async () => {
	await Promise.all([createUserConstraint(), createItemConstraint()]);
};
