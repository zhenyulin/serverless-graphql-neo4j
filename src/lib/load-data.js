import * as Fixtures from '__fixtures__';

import { runCypher, runTransaction } from './neo4j';

export const createUsers = () =>
	runTransaction(
		`CREATE (u:User $user) RETURN u`,
		Fixtures.USERS.map(user => ({ user })),
	);

export const createItems = () =>
	runTransaction(
		`CREATE (i:Item $item) RETURN i`,
		Fixtures.ITEMS.map(item => ({ item })),
	);

export const createUserFollowUsers = () =>
	runTransaction(
		`MATCH (follower:User), (followee:User)
			WHERE follower.id = $followerId AND followee.id = $followeeId
			CREATE (follower)-[r:FOLLOW]->(followee)
			RETURN r`,
		Fixtures.USER_FOLLOW_USERS,
	);

export const createUserLikeItems = () =>
	runTransaction(
		`MATCH (user:User), (item:Item)
			WHERE user.id = $userId AND item.id = $itemId
			CREATE (user)-[r:LIKE]->(item)
			RETURN r`,
		Fixtures.USER_LIKE_ITEMS,
	);

export const loadData = async () => {
	await createUsers();
	await createItems();
	await createUserFollowUsers();
	await createUserLikeItems();
};

export const emptyData = () => runCypher(`MATCH (n) DETACH DELETE n`);
