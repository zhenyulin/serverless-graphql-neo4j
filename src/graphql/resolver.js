import { neo4jgraphql } from 'neo4j-graphql-js';
import { runCypher } from 'lib/neo4j';

export default {
	Query: {
		User: (...args) => neo4jgraphql(...args, false),
		Users: (...args) => neo4jgraphql(...args, false),
		Item: (...args) => neo4jgraphql(...args, false),
		Items: (...args) => neo4jgraphql(...args, false),
	},
	User: {
		followees: ({ id }) =>
			runCypher(
				`
			MATCH (:User{id:${id}})-[:FOLLOW]->(followees:User)
			RETURN followees
		`,
				{},
				true,
			),
		followers: ({ id }) =>
			runCypher(
				`
			MATCH (:User{id:${id}})<-[:FOLLOW]->(followers:User)
			RETURN followers
		`,
				{},
				true,
			),
	},
	Mutation: {
		UserFollowUser: (_, { followerId, followeeId }) =>
			runCypher(`
				MATCH (follower:User {id: "${followerId}"})
				MATCH (followee:User {id: "${followeeId}"})
				MERGE (follower)-[:FOLLOW]->(followee)
				RETURN follower
			`),
		UserUnfollowUser: async (_, { followerId, followeeId }) =>
			runCypher(`
					MATCH (follower:User {id: "${followerId}"})
					MATCH (followee:User {id: "${followeeId}"})
					OPTIONAL MATCH (follower)-[follow:FOLLOW]->(followee)
					DELETE follow
					RETURN follower
				`),
		UserLikeItem: async (_, { userId, itemId }) =>
			runCypher(`
					MATCH (user:User {id: "${userId}"})
					MATCH (item:Item {id: "${itemId}"})
					MERGE (user)-[:LIKE]->(item)
					RETURN user { .*, likeItems: [(user)-[:LIKE]->(items:Item) | items { .* }] } AS user
				`),
		UserDislikeItem: async (_, { userId, itemId }) =>
			runCypher(`
					MATCH (user:User {id: "${userId}"})
					MATCH (item:Item {id: "${itemId}"})
					OPTIONAL MATCH (user)-[like:LIKE]->(item)
					DELETE like
					RETURN user { .*, likeItems: [(user)-[:LIKE]->(items:Item) | items { .* }] } AS user
				`),
	},
};
