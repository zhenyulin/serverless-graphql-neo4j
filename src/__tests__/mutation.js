import { graphql } from 'graphql';

import driver, { runCypher, runCypherOne } from 'lib/neo4j';
import { emptyData, loadData } from 'lib/load-data';
import schema from 'graphql/schema';

const testQuery = async query => {
	const result = await graphql(schema, query);
	expect(result).toMatchSnapshot();
};

describe('Mutation', () => {
	beforeEach(async () => {
		await emptyData();
		await loadData();
	});

	afterAll(async () => {
		driver.close();
	});

	describe('CreateUser', () => {
		it('return the created user', async () => {
			const initState = await runCypherOne(`
				MATCH (u:User { email: "jack.daniel@test.com" })
				RETURN u
			`);
			expect(initState).toBe(undefined);
			await testQuery(`
				mutation {
					CreateUser(user: { name: "Jack Daniel", email: "jack.daniel@test.com" }) {
						name
						email
					}
				}
			`);
			const result = await runCypherOne(`
				MATCH (u:User { email: "jack.daniel@test.com" })
				RETURN u
			`);
			expect(result.email).toBe('jack.daniel@test.com');
		});

		it('throws error if email already existed', async () => {
			await testQuery(`
				mutation {
					CreateUser(user: { name: "Jack Daniel", email: "john.cage@test.com" }) {
						name
						email
					}
				}
			`);
		});

		it('throws error if email is missing', async () => {
			await testQuery(`
				mutation {
					CreateUser(user: { name: "Jack Daniel" }) {
						name
						email
					}
				}
			`);
		});
	});

	describe('UpdateUser', () => {
		it('return the updated user', async () => {
			await testQuery(`
				mutation {
					UpdateUser(id: "4edd40c86762e0fb12000004", user:  { name: "John Lemon" }){
						id
						name
					}
				}
			`);
			const result = await runCypherOne(`
				MATCH (u:User { id: "4edd40c86762e0fb12000004" })
				RETURN u
			`);
			expect(result.name).toBe('John Lemon');
		});
	});

	describe('DeleteUser', () => {
		it('return the deleted user', async () => {
			await testQuery(`
				mutation {
					DeleteUser(id: "4edd40c86762e0fb12000004"){
						id
						name
					}
				}
			`);
			const result = await runCypherOne(`
				MATCH (u:User { id: "4edd40c86762e0fb12000004" })
				RETURN u
			`);
			expect(result).toBe(undefined);
		});
	});

	describe('CreateItem', () => {
		it('return the created item', async () => {
			await testQuery(`
				mutation {
					CreateItem(item: { name: "2046" }){
						name
					}
				}
			`);
			const result = await runCypherOne(`
				MATCH (i:Item { name: "2046" })
				RETURN i
			`);
			expect(result).toMatchSnapshot();
		});

		it('throw error if name is missing', async () => {
			await testQuery(`
				mutation {
					CreateItem(item: {}){
						name
					}
				}
			`);
		});
	});

	describe('UpdateItem', () => {
		it('return the updated item', async () => {
			await testQuery(`
				mutation {
					UpdateItem(id: "4edd40c86762e0fb12000013", item: { name: "Working Class Hero" }){
						id
						name
					}
				}
			`);
			const result = await runCypherOne(`
				MATCH (i:Item { id: "4edd40c86762e0fb12000013" })
				RETURN i
			`);
			expect(result).toMatchSnapshot();
		});
	});

	describe('DeleteItem', () => {
		it('return the deleted item', async () => {
			await testQuery(`
				mutation {
					DeleteItem(id: "4edd40c86762e0fb12000013"){
						id
						name
					}
				}
			`);
			const result = await runCypherOne(`
				MATCH (i:Item { id: "4edd40c86762e0fb12000013" })
				RETURN i
			`);
			expect(result).toBe(undefined);
		});
	});

	describe('UserFollowUser', () => {
		it('return the updated follower', async () => {
			await testQuery(`
				mutation {
					UserFollowUser(followerId: "4edd40c86762e0fb12000005", followeeId: "4edd40c86762e0fb12000004"){
						id
						name
						followees {
							name
						}
					}
				}
			`);
		});

		it('will not create duplicated follow connections', async () => {
			const query = `
				mutation {
					UserFollowUser(followerId: "4edd40c86762e0fb12000005", followeeId: "4edd40c86762e0fb12000004"){
						id
						name
						followees {
							name
						}
					}
				}
			`;
			await graphql(schema, query);
			await graphql(schema, query);
			const result = await runCypher(`
				MATCH (follower:User)-[r:FOLLOW]->(followee:User)
				WHERE follower.id = "4edd40c86762e0fb12000005"
				AND followee.id = "4edd40c86762e0fb12000004"
				RETURN r
			`);
			expect(result).toMatchSnapshot();
		});
	});

	describe('UserUnfollowUser', () => {
		it('return the updated follower', async () => {
			await testQuery(`
				mutation {
					UserUnfollowUser(followerId: "4edd40c86762e0fb12000004", followeeId: "4edd40c86762e0fb12000003"){
						id
						name
						followees {
							name
						}
					}
				}
			`);
			const result = await runCypher(`
				MATCH (follower:User {id: "4edd40c86762e0fb12000004"})
				MATCH (followee:User {id: "4edd40c86762e0fb12000005"})
				MERGE (follower)-[r:FOLLOW]->(followee)
				RETURN r
			`);
			expect(result).toMatchSnapshot();
		});
	});

	describe('UserLikeItem', () => {
		it('return the updated user', async () => {
			await testQuery(`
				mutation {
					UserLikeItem(userId: "4edd40c86762e0fb12000005", itemId: "4edd40c86762e0fb12000013"){
						id
						name
						likeItems {
							name
						}
					}
				}
			`);
		});

		it('will not create duplicated like connections', async () => {
			const query = `
				mutation {
					UserLikeItem(userId: "4edd40c86762e0fb12000005", itemId: "4edd40c86762e0fb12000013"){
						id
						name
						likeItems {
							name
						}
					}
				}
			`;
			await graphql(schema, query);
			await graphql(schema, query);
			const result = await runCypher(`
				MATCH (user:User)-[r:LIKE]->(item:Item)
				WHERE user.id = "4edd40c86762e0fb12000005"
				AND item.id = "4edd40c86762e0fb12000013"
				RETURN r
			`);
			expect(result).toMatchSnapshot();
		});
	});

	describe('UserDislikeItem', () => {
		it('return the updated user', async () => {
			await testQuery(`
				mutation {
					UserDislikeItem(userId: "4edd40c86762e0fb12000004", itemId: "4edd40c86762e0fb12000014"){
						id
						name
						likeItems {
							name
						}
					}
				}
			`);
			const result = await runCypher(`
				MATCH (user:User)-[r:LIKE]->(item:Item)
				WHERE user.id = "4edd40c86762e0fb12000004"
				AND item.id = "4edd40c86762e0fb12000014"
				RETURN r
			`);
			expect(result).toMatchSnapshot();
		});
	});
});
