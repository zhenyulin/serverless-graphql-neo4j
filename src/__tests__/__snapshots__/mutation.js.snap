// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation CreateItem return the created item 1`] = `
Object {
  "data": Object {
    "CreateItem": Object {
      "name": "2046",
    },
  },
}
`;

exports[`Mutation CreateItem return the created item 2`] = `
Object {
  "name": "2046",
}
`;

exports[`Mutation CreateItem throw error if name is missing 1`] = `
Object {
  "data": Object {
    "CreateItem": null,
  },
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field Item.name.],
  ],
}
`;

exports[`Mutation CreateUser return the created user 1`] = `
Object {
  "data": Object {
    "CreateUser": Object {
      "email": "jack.daniel@test.com",
      "name": "Jack Daniel",
    },
  },
}
`;

exports[`Mutation CreateUser throws error if email already existed 1`] = `
Object {
  "data": Object {
    "CreateUser": Object {
      "email": "john.cage@test.com",
      "name": "Jack Daniel",
    },
  },
}
`;

exports[`Mutation CreateUser throws error if email is missing 1`] = `
Object {
  "data": Object {
    "CreateUser": null,
  },
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field User.email.],
  ],
}
`;

exports[`Mutation DeleteItem return the deleted item 1`] = `
Object {
  "data": Object {
    "DeleteItem": null,
  },
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field Item.id.],
  ],
}
`;

exports[`Mutation DeleteUser return the deleted user 1`] = `
Object {
  "data": Object {
    "DeleteUser": null,
  },
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field User.id.],
  ],
}
`;

exports[`Mutation UpdateItem return the updated item 1`] = `
Object {
  "data": Object {
    "UpdateItem": Object {
      "id": "4edd40c86762e0fb12000013",
      "name": "Working Class Hero",
    },
  },
}
`;

exports[`Mutation UpdateItem return the updated item 2`] = `
Object {
  "id": "4edd40c86762e0fb12000013",
  "name": "Working Class Hero",
}
`;

exports[`Mutation UpdateUser return the updated user 1`] = `
Object {
  "data": Object {
    "UpdateUser": Object {
      "id": "4edd40c86762e0fb12000004",
      "name": "John Lemon",
    },
  },
}
`;

exports[`Mutation UserDislikeItem return the updated user 1`] = `
Object {
  "data": Object {
    "UserDislikeItem": Object {
      "id": "4edd40c86762e0fb12000004",
      "likeItems": Array [],
      "name": "John Lennon",
    },
  },
}
`;

exports[`Mutation UserDislikeItem return the updated user 2`] = `Array []`;

exports[`Mutation UserFollowUser return the updated follower 1`] = `
Object {
  "data": Object {
    "UserFollowUser": Object {
      "followees": Array [
        Object {
          "name": "John Lennon",
        },
      ],
      "id": "4edd40c86762e0fb12000005",
      "name": "Miles Davis",
    },
  },
}
`;

exports[`Mutation UserFollowUser will not create duplicated follow connections 1`] = `
Array [
  Object {},
]
`;

exports[`Mutation UserLikeItem return the updated user 1`] = `
Object {
  "data": Object {
    "UserLikeItem": Object {
      "id": "4edd40c86762e0fb12000005",
      "likeItems": Array [
        Object {
          "name": "Imagine",
        },
      ],
      "name": "Miles Davis",
    },
  },
}
`;

exports[`Mutation UserLikeItem will not create duplicated like connections 1`] = `
Array [
  Object {},
]
`;

exports[`Mutation UserUnfollowUser return the updated follower 1`] = `
Object {
  "data": Object {
    "UserUnfollowUser": Object {
      "followees": Array [],
      "id": "4edd40c86762e0fb12000004",
      "name": "John Lennon",
    },
  },
}
`;

exports[`Mutation UserUnfollowUser return the updated follower 2`] = `
Array [
  Object {},
]
`;
